[core] Added SRT time API (#1367)
Haivision/srt: docs/API-functions.md

   int flags;            // Left for future
   int msgttl;           // TTL for a message, default -1 (no TTL limitation)
   int inorder;          // Whether a message is allowed to supersede partially lost one. Unused in stream and live mode.
   int boundary;         // 0:mid pkt, 1(01b):end of frame, 2(11b):complete frame, 3(10b): start of frame

>    int64_t srctime;      // source time in mic(usec), 0: use internal time     

Reviewer: What is "mic"? I think it should say something like: "source time (usec since steady clock epoch)". It is extremely important that users do not specify here the time read from `gettimeofday` because it make work for them most of the time making a false impression that their code is correct.

Author: Fixed

The following flags are available, as collected in `logging_api.h` public header:

- `SRT_LOGF_DISABLE_TIME`: Do not provide the time in the header
- `SRT_LOGF_DISABLE_THREADNAME`: Do not provide the thread name in the header
- `SRT_LOGF_DISABLE_SEVERITY`: Do not provide severity information in the header
- `SRT_LOGF_DISABLE_EOL`: Do not add the end-of-line character to the log line

## Time Access

The following set of functions is intended to retrieve timestamps from clock used by SRT.
This timestamp MUST be used by sender if it wants to explicitely assign a timestamp
to a certain packet. The sender can pass the timestamp in `MSGCTRL::srctime` of `srt_sendmsg2(..)`
function together with the packet being submitted to SRT.
The time provided MUST equal or exceed connection start time (`srt_connection_uptime(..)`)
of the SRT socket passed to `srt_sendmsg2(..)`.

> The current time value as of SRT internal clock can be retrieved using `srt_time_now()` function.

Reviewer: Consider placing the following information:

There are two known use cases of why you might want to use `srctime`:

1. SRT passthrough (for stream gateways). In this case you simply retrieve packets from an SRT source and pass them transparently to an SRT output (possibly re-encrypting). In that case, for every packet you read should also preserve the original value of `srctime` as obtained from `srt_recvmsg2`, and the original `srctime` for the same packet should be then passed for `srt_sendmsg2`. This could be used to avoid time jitter resulting from varying time difference between the time of receiving and sending the same packet.

2. Stable timing source. Your live streaming procedure, when spreading packets evenly into the stream in time, might predefine times for every single packet to keep time distances perfectly equal, or you believe that your signal input delivers packets at exact times that should be assigned to them - and you would like to preserve these times at the SRT receiving side and avoid a jitter that may result from varying time differences between the packet arrival and the moment when sending it over SRT. In cases like that you might do the following:

* When the packet arrival time counts, grab current time at this moment using `srt_time_now()`.
* When you wanted a precalculated packet time, use your private relative time counter reset to 0 when your first packet is ready, and specify `srctime` as a sum of this counter and the connection uptime obtained from `srt_connection_uptime()`. Although you still have to synchronize sending times with these predefined times, by explicitly specifying the source time you avoid the jitter resulting from a lost accuracy of waiting time and unfortunate thread scheduling.

Note that the clock used for `srctime` is a "steady clock" provided by system libraries and **it is not the clock expressing the current system time** (do not use functions like `ftime`, `time` or `gettimeofday` for it). This is a monotonic clock, that is, the base point in time when this time has a value of 0 is unchangeable during the whole lifetime of the application. An example of such clock is `CLOCK_MONOTONIC` used by `clock_gettime` function or `std::chrono::steady_clock` in the standard C++ library. To avoid problems and misunderstanding you should rely exclusively on time values provided by `srt_time_now()` and `srt_connection_uptime()` functions.

The following flags are available, as collected in `logging_api.h` public header:

- `SRT_LOGF_DISABLE_TIME`: Do not provide the time in the header
- `SRT_LOGF_DISABLE_THREADNAME`: Do not provide the thread name in the header
- `SRT_LOGF_DISABLE_SEVERITY`: Do not provide severity information in the header
- `SRT_LOGF_DISABLE_EOL`: Do not add the end-of-line character to the log line

## Time Access

> The following set of functions is intended to retrieve timestamps from clock used by SRT.
> This timestamp MUST be used by sender if it wants to explicitely assign a timestamp
> to a certain packet. The sender can pass the timestamp in `MSGCTRL::srctime` of `srt_sendmsg2(..)`
> function together with the packet being submitted to SRT.
> The time provided MUST equal or exceed connection start time (`srt_connection_time(..)`)
> of the SRT socket passed to `srt_sendmsg2(..)`.
> 
> The current time value as of SRT internal clock can be retrieved using `srt_time_now()` function.

Reviewer 2: ```suggestion
The following set of functions is intended to retrieve timestamps from the clock used by SRT.
Such a timestamp MUST be used by the sender to explicitly assign a timestamp
to a certain packet. The sender can pass the timestamp in `MSGCTRL::srctime` of the `srt_sendmsg2(..)`
function together with the packet being submitted to SRT.
The time value provided MUST equal or exceed the connection start time (`srt_connection_time(..)`)
of the SRT socket passed to `srt_sendmsg2(..)`.

The current time value of the SRT internal clock can be retrieved using the `srt_time_now()` function.
```

Reviewer: The second sentence: there's no obligation to use the timestamp (the default will be used otherwise), the conditional statement here is important. I don't know how to best express it.

What is important here is that if the sender procedure wants to explicitly specify the timestamp, it must be either the timestamp read from the SRT source when reading a packet (and it's provided in the `SRT_MSGCTRL::srctime` field when reading the packet), or a calculated time not earlier than the connection start time (obtained from `srt_connection_time()`).

Author: Maybe this way then:

The sender can pass the timestamp in `MSGCTRL::srctime` of the `srt_sendmsg2(..)`
function together with the packet being submitted to SRT.
If the `srctime` value is not provided (the default value of 0 is set), SRT will use internal
clock and assign the packet submission time as the packet timestamp.
If the sender wants to explicitly assign a timestamp
to a certain packet. this timestamp MUST be taken from SRT Time Access functions.
The time value provided MUST equal or exceed the connection start time (`srt_connection_time(..)`)
of the SRT socket passed to `srt_sendmsg2(..)`.

The current time value as of the SRT internal clock can be retrieved using the `srt_time_now()` function.

> There are two known use cases of why you might want to use `srctime`:

Reviewer 2: ```suggestion
There are two known cases where you might want to use `srctime`:
```

Reviewer: The term "use cases" was there for a reason. Maybe "There are two known use cases for `srctime`"?

Author: I think "cases" are also good.

Reviewer 2: The word "use" in the same sentence qualifies "case".

1. SRT passthrough (for stream gateways).

> In this case you simply retrieve packets from an SRT source and pass them transparently
> to an SRT output (possibly re-encrypting). In that case, for every packet you read
> should also preserve the original value of `srctime` as obtained from `srt_recvmsg2`,
> and the original `srctime` for the same packet should be then passed for `srt_sendmsg2`.
> This could be used to avoid time jitter resulting from varying time difference between
> the time of receiving and sending the same packet.

Reviewer 2: ```suggestion
You may wish to simply retrieve packets from an SRT source and pass them transparently
to an SRT output (possibly re-encrypting). In that case, every packet you read
should preserve the original value of `srctime` as obtained from `srt_recvmsg2`,
and the original `srctime` for each packet should be then passed to `srt_sendmsg2`.
This mechanism could be used to avoid jitter resulting from varying differences between
the time of receiving and sending the same packet.
```

2. Stable timing source.

> Your live streaming procedure, when spreading packets evenly into the stream in time,
> might predefine times for every single packet to keep time distances perfectly equal,
> or you believe that your signal input delivers packets at exact times that should be
> assigned to them - and you would like to preserve these times at the SRT receiving side
> and avoid a jitter that may result from varying time differences between the packet arrival
> and the moment when sending it over SRT. In cases like that you might do the following:

Reviewer 2: ```suggestion
In the case of a live streaming procedure, when spreading packets evenly into the stream,
you might want to predefine times for every single packet to keep time intervals perfectly equal.
Or, if you believe that your input signal delivers packets at the exact times that should be
assigned to them, you might want to preserve these times at the SRT receiving side
to avoid jitter that may result from varying time differences between the packet arrival
and the moment when sending it over SRT. In such cases you might do the following:
```

>   - When the packet arrival time counts, grab current time at this moment using `srt_time_now()`.
> 
>   - When you want a precalculated packet time, use your private relative time counter
>   set at the moment when the connection was made, then since the moment when your first packet
>   is ready, start precalculating packet times relatively to the connection start time obtained
>   from `srt_connection_time()`. Although you still have to synchronize sending times with these
>   predefined times, by explicitly specifying the source time you avoid the jitter
>   resulting from a lost accuracy of waiting time and unfortunate thread scheduling.

Reviewer 2: ```suggestion
  - At the packet arrival time, grab the current time at that moment using `srt_time_now()`.

  - When you want a precalculated packet time, use a private relative time counter
  set at the moment when the connection was made. From the moment when your first packet
  is ready, start precalculating packet times relative to the connection start time obtained
  from `srt_connection_time()`. Although you still have to synchronize sending times with these
  predefined times, by explicitly specifying the source time you avoid the jitter
  resulting from a lost accuracy due to waiting time and unfortunate thread scheduling.
```

> Note that the time used for `srctime` is using an internally defined clock

Reviewer 2: ```suggestion
Note that  `srctime` uses an internally defined clock
```
